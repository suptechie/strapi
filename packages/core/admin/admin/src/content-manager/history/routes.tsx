/* eslint-disable check-file/filename-naming-convention */
import { useLocation, type RouteObject, matchRoutes } from 'react-router-dom';

/**
 * These routes will be merged with the rest of the Content Manager routes
 */
const routes: RouteObject[] = [
  {
    path: ':collectionType/:slug/:id/history',
    lazy: async () => {
      const { HistoryPage } = await import('./pages/History');

      return {
        Component: HistoryPage,
      };
    },
  },
  {
    path: ':collectionType/:slug/history',
    lazy: async () => {
      const { HistoryPage } = await import('./pages/History');

      return {
        Component: HistoryPage,
      };
    },
  },
];

/**
 * Used to determine if we're on a history route from the admin and the content manager,
 * so that we can hide the left menus on all history routes
 */
function useIsHistoryRoute() {
  const location = useLocation();
  const historyRoutes = routes.map((route) => ({ path: `content-manager/${route.path}` }));
  const matches = matchRoutes(historyRoutes, location);

  return Boolean(matches);
}

export { routes, useIsHistoryRoute };
