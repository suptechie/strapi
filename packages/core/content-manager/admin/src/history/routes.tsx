/* eslint-disable check-file/filename-naming-convention */
import { lazy } from 'react';

import { type PathRouteProps } from 'react-router-dom';

const ProtectedHistoryPage = lazy(() =>
  import('./pages/History').then((mod) => ({ default: mod.ProtectedHistoryPage }))
);

/**
 * These routes will be merged with the rest of the Content Manager routes
 */
const routes: PathRouteProps[] = [
  {
    path: ':collectionType/:slug/:id/history',
    Component: ProtectedHistoryPage,
  },
  {
    path: ':collectionType/:slug/history',
    Component: ProtectedHistoryPage,
  },
];

export { routes };
