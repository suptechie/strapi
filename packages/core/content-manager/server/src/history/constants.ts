export const HISTORY_VERSION_UID = 'plugin::content-manager.history-version';
export const FIELDS_TO_IGNORE = [
  'createdAt',
  'updatedAt',
  'publishedAt',
  'createdBy',
  'updatedBy',
  'locale',
  'strapi_stage',
  'strapi_assignee',
];
