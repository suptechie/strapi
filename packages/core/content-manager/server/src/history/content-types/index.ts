import { historyVersion } from './history-version';

export const contentTypes = {
  'history-version': historyVersion,
};
