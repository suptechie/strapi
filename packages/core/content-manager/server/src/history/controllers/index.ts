import type { Plugin } from '@strapi/types';

/**
 * The controllers will me merged with the other Content Manager controllers,
 * so we need to avoid conficts in the controller names.
 */
export const controllers: Plugin.LoadedPlugin['controllers'] = {};
