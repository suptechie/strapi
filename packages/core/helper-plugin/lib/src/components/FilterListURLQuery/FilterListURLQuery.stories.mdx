<!--- FilterListURLQuery.stories.mdx --->

import { useEffect } from 'react';
import { Meta, ArgsTable, Canvas, Story } from '@storybook/addon-docs';
import { Button, Box, Main, Row, Stack } from '@strapi/parts';
import useQueryParams from '../../hooks/useQueryParams';
import FilterListURLQuery from './index';

<Meta title="components/FilterListURLQuery" />

# FilterListURLQuery

This component is used to change the page size.

## Imports

```js
import { FilterListURLQuery } from '@strapi/helper-plugin';
```

## Usage

<Canvas>
  <Story name="base">
    {() => {
      const [{ query }, setQuery] = useQueryParams();
      const filters = {
        $and: [{ city: { $eq: 'paris' } }, { categories: { name: { $eq: 'paris' } } }],
      };
      useEffect(() => {
        setQuery({ filters });
      }, []);
      const handleResetFilters = () => setQuery({ filters });
      return (
        <Main>
          <Stack size={6}>
            <Row paddingTop={6}>
              <FilterListURLQuery
                filtersSchema={[
                  {
                    name: 'categories',
                    fieldSchema: {
                      type: 'relation',
                      mainField: { name: 'name', schema: { type: 'string' } },
                    },
                    metadatas: { label: 'categories' },
                  },
                  {
                    name: 'city',
                    fieldSchema: {
                      type: 'string',
                    },
                    metadatas: { label: 'city' },
                  },
                ]}
              />
            </Row>
            <Box>
              <Button onClick={handleResetFilters}>Reset filters</Button>
            </Box>
          </Stack>
        </Main>
      );
    }}
  </Story>
</Canvas>

### Props

<ArgsTable of={FilterListURLQuery} />
