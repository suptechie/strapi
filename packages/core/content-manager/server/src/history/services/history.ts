import type { Core, Data, Schema, Struct } from '@strapi/types';
import { async, errors } from '@strapi/utils';
import { omit } from 'lodash/fp';

import { FIELDS_TO_IGNORE, HISTORY_VERSION_UID } from '../constants';
import type { HistoryVersions } from '../../../../shared/contracts';
import {
  CreateHistoryVersion,
  HistoryVersionDataResponse,
} from '../../../../shared/contracts/history-versions';
import { createServiceUtils } from './utils';
import { getService as getContentManagerService } from '../../utils';

// Needed because the query engine doesn't return any types yet
type HistoryVersionQueryResult = Omit<HistoryVersionDataResponse, 'locale'> &
  Pick<CreateHistoryVersion, 'locale'>;

const createHistoryService = ({ strapi }: { strapi: Core.Strapi }) => {
  const query = strapi.db.query(HISTORY_VERSION_UID);
  const serviceUtils = createServiceUtils({ strapi });

  return {
    async createVersion(historyVersionData: HistoryVersions.CreateHistoryVersion) {
      await query.create({
        data: {
          ...historyVersionData,
          createdAt: new Date(),
          createdBy: strapi.requestContext.get()?.state?.user.id,
        },
      });
    },

    async findVersionsPage(params: HistoryVersions.GetHistoryVersions.Request): Promise<{
      results: HistoryVersions.HistoryVersionDataResponse[];
      pagination: HistoryVersions.Pagination;
    }> {
      const locale = params.query.locale || (await serviceUtils.getDefaultLocale());
      const [{ results, pagination }, localeDictionary] = await Promise.all([
        query.findPage({
          ...params.query,
          where: {
            $and: [
              { contentType: params.query.contentType },
              ...(params.query.documentId ? [{ relatedDocumentId: params.query.documentId }] : []),
              ...(locale ? [{ locale }] : []),
            ],
          },
          populate: ['createdBy'],
          orderBy: [{ createdAt: 'desc' }],
        }),
        serviceUtils.getLocaleDictionary(),
      ]);
      const populateEntryRelations = async (
        entry: HistoryVersionQueryResult
      ): Promise<CreateHistoryVersion['data']> => {
        const entryWithRelations = await Object.entries(entry.schema).reduce(
          async (currentDataWithRelations, [attributeKey, attributeSchema]) => {
            const attributeValue = entry.data[attributeKey];
            const attributeValues = Array.isArray(attributeValue)
              ? attributeValue
              : [attributeValue];

            if (attributeSchema.type === 'media') {
              const permissionChecker = getContentManagerService('permission-checker').create({
                userAbility: params.state.userAbility,
                model: 'plugin::upload.file',
              });

              const response = await serviceUtils.buildMediaResponse(attributeValues);
              const sanitizedResults = await Promise.all(
                response.results.map((media) => permissionChecker.sanitizeOutput(media))
              );

              return {
                ...(await currentDataWithRelations),
                [attributeKey]: {
                  results: sanitizedResults,
                  meta: response.meta,
                },
              };
            }

            // TODO: handle relations that are inside components
            if (
              attributeSchema.type === 'relation' &&
              attributeSchema.relation !== 'morphToOne' &&
              attributeSchema.relation !== 'morphToMany'
            ) {
              /**
               * Don't build the relations response object for relations to admin users,
               * because pickAllowedAdminUserFields will sanitize the data in the controller.
               */
              if (attributeSchema.target === 'admin::user') {
                const adminUsers = await Promise.all(
                  attributeValues.map((userToPopulate) => {
                    if (userToPopulate == null) {
                      return null;
                    }

                    return strapi
                      .query('admin::user')
                      .findOne({ where: { id: userToPopulate.id } });
                  })
                );

                return {
                  ...(await currentDataWithRelations),
                  /**
                   * Ideally we would return the same "{results: [], meta: {}}" shape, however,
                   * when sanitizing the data as a whole in the controller before sending to the client,
                   * the data for admin relation user is completely sanitized if we return an object here as opposed to an array.
                   */
                  [attributeKey]: adminUsers,
                };
              }

              const permissionChecker = getContentManagerService('permission-checker').create({
                userAbility: params.state.userAbility,
                model: attributeSchema.target,
              });

              const response = await serviceUtils.buildRelationReponse(
                attributeValues,
                attributeSchema
              );
              const sanitizedResults = await Promise.all(
                response.results.map((media) => permissionChecker.sanitizeOutput(media))
              );

              return {
                ...(await currentDataWithRelations),
                [attributeKey]: {
                  results: sanitizedResults,
                  meta: response.meta,
                },
              };
            }

            // Not a media or relation, nothing to change
            return currentDataWithRelations;
          },
          Promise.resolve(entry.data)
        );

        return entryWithRelations;
      };

      const formattedResults = await Promise.all(
        (results as HistoryVersionQueryResult[]).map(async (result) => {
          return {
            ...result,
            data: await populateEntryRelations(result),
            meta: {
              unknownAttributes: serviceUtils.getSchemaAttributesDiff(
                result.schema,
                strapi.getModel(params.query.contentType).attributes
              ),
            },
            locale: result.locale ? localeDictionary[result.locale] : null,
          };
        })
      );

      return {
        results: formattedResults,
        pagination,
      };
    },

    async restoreVersion(versionId: Data.ID) {
      const version = await query.findOne({ where: { id: versionId } });
      const contentTypeSchemaAttributes = strapi.getModel(version.contentType).attributes;
      const schemaDiff = serviceUtils.getSchemaAttributesDiff(
        version.schema,
        contentTypeSchemaAttributes
      );

      // Set all added attribute values to null
      const dataWithoutAddedAttributes = Object.keys(schemaDiff.added).reduce(
        (currentData, addedKey) => {
          currentData[addedKey] = null;
          return currentData;
        },
        // Clone to avoid mutating the original version data
        structuredClone(version.data)
      );
      const sanitizedSchemaAttributes = omit(
        FIELDS_TO_IGNORE,
        contentTypeSchemaAttributes
      ) as Struct.SchemaAttributes;

      // Set all deleted relation values to null
      const reducer = async.reduce(Object.entries(sanitizedSchemaAttributes));
      const dataWithoutMissingRelations = await reducer(
        async (
          previousRelationAttributes: Record<string, unknown>,
          [name, attribute]: [string, Schema.Attribute.AnyAttribute]
        ) => {
          const versionRelationData = version.data[name];
          if (!versionRelationData) {
            return previousRelationAttributes;
          }

          if (
            attribute.type === 'relation' &&
            // TODO: handle polymorphic relations
            attribute.relation !== 'morphToOne' &&
            attribute.relation !== 'morphToMany'
          ) {
            const data = await serviceUtils.getRelationRestoreValue(versionRelationData, attribute);
            previousRelationAttributes[name] = data;
          }

          if (attribute.type === 'media') {
            const data = await serviceUtils.getMediaRestoreValue(versionRelationData, attribute);
            previousRelationAttributes[name] = data;
          }

          return previousRelationAttributes;
        },
        // Clone to avoid mutating the original version data
        structuredClone(dataWithoutAddedAttributes)
      );

      const data = omit(['id', ...Object.keys(schemaDiff.removed)], dataWithoutMissingRelations);
      const restoredDocument = await strapi.documents(version.contentType).update({
        documentId: version.relatedDocumentId,
        locale: version.locale,
        data,
      });

      if (!restoredDocument) {
        throw new errors.ApplicationError('Failed to restore version');
      }

      return restoredDocument;
    },
  };
};

export { createHistoryService };
