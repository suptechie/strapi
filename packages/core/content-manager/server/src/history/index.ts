import type { Plugin } from '@strapi/types';
import { controllers } from './controllers';
import { services } from './services';
import { routes } from './routes';
import { getService } from './utils';
import { historyVersion } from './models/history-version';

/**
 * Check once if the feature is enabled (both license info & feature flag) before loading it,
 * so that we can assume it is enabled in the other files.
 */
const getFeature = (): Partial<Plugin.LoadedPlugin> => {
  // TODO: add license check here when it's ready on the license registry
  if (strapi.features.future.isEnabled('history')) {
    return {
      register({ strapi }) {
        strapi.get('models').add(historyVersion);
      },
      bootstrap({ strapi }) {
        // Start recording history and saving history versions
        getService(strapi, 'history').init();
      },
      controllers,
      services,
      routes,
    };
  }

  /**
   * Keep registering the model to avoid losing the data if the feature is disabled,
   * or if the license expires.
   */
  return {
    register({ strapi }) {
      strapi.get('models').add(historyVersion);
    },
  };
};

export default getFeature();
