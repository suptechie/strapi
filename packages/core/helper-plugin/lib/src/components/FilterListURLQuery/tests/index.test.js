/**
 *
 * Tests for FilterListURLQuery
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { ThemeProvider, lightTheme } from '@strapi/parts';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import FilterListURLQuery from '../index';

const makeApp = (history, filtersSchema) => (
  <Router history={history}>
    <ThemeProvider theme={lightTheme}>
      <IntlProvider
        locale="en"
        messages={{
          'components.FilterOptions.FILTER_TYPES.$eq': 'is',
          'components.FilterOptions.FILTER_TYPES.$contains': 'contains',
        }}
        defaultLocale="en"
        textComponent="span"
      >
        <FilterListURLQuery filtersSchema={filtersSchema} />
      </IntlProvider>
    </ThemeProvider>
  </Router>
);

describe('<FilterListURLQuery />', () => {
  it('renders and matches the snapshot', () => {
    const history = createMemoryHistory();
    const filtersSchema = [
      {
        name: 'bool',
        metadatas: { label: 'bool' },
        fieldSchema: { type: 'boolean' },
      },
      {
        name: 'date',
        metadatas: { label: 'date' },
        fieldSchema: { type: 'date' },
      },
      {
        name: 'enum',
        metadatas: { label: 'enum' },
        fieldSchema: { type: 'enumeration', options: ['one', 'two', 'three'] },
      },
      {
        name: 'long',
        metadatas: { label: 'long' },
        fieldSchema: { type: 'string' },
      },
      {
        name: 'many_to_one',
        metadatas: { label: 'many_to_one' },
        fieldSchema: {
          type: 'relation',
          mainField: {
            name: 'postal_code',
            schema: { type: 'string', pluginOptions: { i18n: { localized: true } } },
          },
        },
      },
      {
        name: 'number',
        metadatas: { label: 'number' },
        fieldSchema: { type: 'integer' },
      },
      {
        name: 'time',
        metadatas: { label: 'time' },
        fieldSchema: { type: 'time' },
      },
    ];
    history.push({
      pathname: '/',
      search:
        'sort=city&filters[$and][0][bool][$eq]=true&filters[$and][1][date][$eq]=2021-09-01&filters[$and][2][enum][$eq]=one&filters[$and][3][long][$eq]=test&filters[$and][4][many_to_one][postal_code][$eq]=test&filters[$and][5][number][$eq]=2&filters[$and][6][time][$contains]=00:45',
    });
    const { container } = render(makeApp(history, filtersSchema));

    expect(container).toMatchSnapshot();
  });
});
