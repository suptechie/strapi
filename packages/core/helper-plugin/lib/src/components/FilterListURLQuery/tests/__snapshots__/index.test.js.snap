// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterListURLQuery /> renders and matches the snapshot 1`] = `
.c7 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 {
  padding: 4px;
}

.c4 {
  font-weight: 500;
  font-size: 0.75rem;
  line-height: 1.33;
  color: #32324d;
}

.c1 {
  background: #f0f0ff;
  color: #4945ff;
  padding-right: 12px;
  padding-left: 12px;
}

.c6 {
  padding-left: 8px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  border: 1px solid #d9d8ff;
  border-radius: 4px;
  height: 2rem;
}

.c2 svg {
  height: 0.5rem;
  width: 0.5rem;
}

.c2 svg path {
  fill: #4945ff;
}

.c5 {
  border-right: 1px solid #d9d8ff;
  padding-right: 8px;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          bool
           
          is
           
          true
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#32324D"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          date
           
          is
           
          Wednesday, September 1, 2021
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#32324D"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          enum
           
          is
           
          one
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#32324D"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          long
           
          is
           
          test
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#32324D"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          many_to_one
           
          is
           
          test
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#32324D"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          number
           
          is
           
          2
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#32324D"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c0"
  >
    <button
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4 c5"
        >
          time
           
          contains
           
          12:45 AM
        </span>
        <div
          class="c6"
        >
          <div
            class="c3"
          >
            <svg
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 2.417L21.583 0 12 9.583 2.417 0 0 2.417 9.583 12 0 21.583 2.417 24 12 14.417 21.583 24 24 21.583 14.417 12 24 2.417z"
                fill="#32324D"
              />
            </svg>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div
    class="c7"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
